src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2666,38 +2666,87 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
-        // This code only handles the most common formatting sequences, which are
-        // all of the pattern:
-        //     ESC '[' digits+ (';' digits)* 'm'
+        // This code only handles the following formatting sequences, corresponding to
+        // the patterns:
+        //     CSI params? 'm'               // SGR: Select Graphics Rendition
+        //     CSI params? '#' [{}pq]        // XTPUSHSGR ('{'), XTPOPSGR ('}'), or their aliases ('p' and 'q')
         //
-        // There are many other VT100 escape sequences, but this simple pattern
-        // is sufficient for our formatting system.  We won't handle cursor movements
-        // or other attempts at animation.
+        // Where:
+        //     CSI:    ESC '[' OR \x009b     // (C0 or C1 CSI)
+        //     params: digit+ (';' params)?
         //
-        // Note that offset is adjusted past the escape sequence.
+        // There are many other VT100 escape sequences, but these text attribute sequences
+        // (color-related, underline, etc.) are sufficient for our formatting system.  We
+        // won't handle cursor movements or other attempts at animation.
+        //
+        // Note that offset is adjusted past the escape sequence, or at least one
+        // character forward if there is no escape sequence at the specified position.
         internal static int ControlSequenceLength(string str, ref int offset)
         {
             var start = offset;
-            if (str[offset++] != (char)0x1B)
+
+            // First, check for the CSI:
+            if ((str[offset] == (char)0x1b) && (str.Length > (offset + 1)) && (str[offset + 1] == '['))
+            {
+                // C0 CSI
+                offset += 2;
+            }
+            else if (str[offset] == (char)0x9b)
+            {
+                // C1 CSI
+                offset += 1;
+            }
+            else
+            {
+                // We need to skip over at least the current character.
+                offset += 1;
                 return 0;
+            }
 
-            if (offset >= str.Length || str[offset] != '[')
+            if (offset >= str.Length)
+            {
                 return 0;
+            }
 
-            offset += 1;
-            while (offset < str.Length)
+            // Next, handle possible numeric arguments:
+            char c;
+            do
             {
-                var c = str[offset++];
-                if (c == 'm')
-                    break;
+                c = str[offset++];
+            }
+            while ((offset < str.Length) && (char.IsDigit(c) || c == ';'));
 
-                if (char.IsDigit(c) || c == ';')
-                    continue;
+            // Finally, handle the command characters for the specific sequences we
+            // handle:
+            if (c == 'm')
+            {
+                // SGR: Select Graphics Rendition
+                return offset - start;
+            }
+            else if (c == '#')
```

Comment:
We could remove "else" for more readability.
```suggestion

            if (c == '#')
```